src/evict/evict_lru.c

```diff
@@ -1987,6 +1995,230 @@ __evict_walk_tree(WT_SESSION_IMPL *session, WT_EVICT_QUEUE *queue, u_int max_ent
+        if (__wt_page_evict_urgent(session, ref))
+            *urgent_queuedp = true;
+        return;
+    }
+
+    /* Pages that are empty or from dead trees are fast-tracked. */
+    if (__wt_page_is_empty(page) || F_ISSET(session->dhandle, WT_DHANDLE_DEAD))
+        goto fast;
+
+    /* Skip pages we don't want. */
+    want_page = (F_ISSET(cache, WT_CACHE_EVICT_CLEAN) && !modified) ||
+      (F_ISSET(cache, WT_CACHE_EVICT_DIRTY) && modified) ||
+      (F_ISSET(cache, WT_CACHE_EVICT_UPDATES) && page->modify != NULL);
+    if (!want_page) {
+        WT_STAT_CONN_INCR(session, cache_eviction_server_skip_unwanted_pages);
+        return;
+    }
+
+    /*
+     * Do not evict a clean metadata page that contains historical data needed to satisfy a reader.
+     * Since there is no history store for metadata, we won't be able to serve an older reader if we
+     * evict this page.
+     */
+    if (WT_IS_METADATA(session->dhandle) && F_ISSET(cache, WT_CACHE_EVICT_CLEAN_HARD) &&
+      F_ISSET(ref, WT_REF_FLAG_LEAF) && !modified && page->modify != NULL &&
+      !__wt_txn_visible_all(session, page->modify->rec_max_txn, page->modify->rec_max_timestamp)) {
+        WT_STAT_CONN_INCR(session, cache_eviction_server_skip_metatdata_with_history);
+        return;
+    }
+
+    /*
+     * Don't attempt eviction of internal pages with children in cache (indicated by seeing an
+     * internal page that is the parent of the last page we saw).
+     *
+     * Also skip internal page unless we get aggressive, the tree is idle (indicated by the tree
+     * being skipped for walks), or we are in eviction debug mode. The goal here is that if trees
+     * become completely idle, we eventually push them out of cache completely.
+     */
+    if (!FLD_ISSET(conn->debug_flags, WT_CONN_DEBUG_EVICT_AGGRESSIVE_MODE) &&
+      F_ISSET(ref, WT_REF_FLAG_INTERNAL)) {
+        if (page == last_parent) {
+            WT_STAT_CONN_INCR(session, cache_eviction_server_skip_intl_page_with_active_child);
+            return;
+        }
+        if (__wt_atomic_load32(&btree->evict_walk_period) == 0 && !__wt_cache_aggressive(session))
+            return;
+    }
+
+    /* Evaluate dirty page candidacy, when eviction is not aggressive. */
+    if (!__wt_cache_aggressive(session) && modified && __evict_skip_dirty_candidate(session, page))
+        return;
+
+fast:
+    /* If the page can't be evicted, give up. */
+    if (!__wt_page_can_evict(session, ref, NULL))
+        return;
+
+    WT_ASSERT(session, evict->ref == NULL);
+    if (!__evict_push_candidate(session, queue, evict, ref))
+        return;
+
+    *queuedp = true;
+    __wt_verbose(session, WT_VERB_EVICTSERVER, "walk select: %p, size %" WT_SIZET_FMT, (void *)page,
+      __wt_atomic_loadsize(&page->memory_footprint));
+
+    return;
+}
+
+/*
+ * __evict_walk_tree --
+ *     Get a few page eviction candidates from a single underlying file.
+ */
```

Comment:
Just a note: Let's make sure this change gets backported to **all** older versions so that fixes are easier to apply.